### ISSUE TITLE
Make UNNEST with JOIN clearer in the documentation.

### ISSUE THREAD

**Issue by bitsondatadev** [trinodb/trino#17897]
Taken from this awesome example from @kasiafi, I was initially confused when I ran the following query, I got the error message, `Column 't2.y' cannot be resolved`:

```
WITH
  t (x, y) AS (
    VALUES
      (1, ARRAY['a', 'b', 'c'])
  ),
  t2 (x, y) AS (
    VALUES
      (1, ARRAY['a', 'b', 'c'])
  ),
  t3 (x, y) AS (
    SELECT
      x,
      y_unnested
    FROM
      t
      CROSS JOIN UNNEST (t2.y) t4 (y_unnested)
  )
SELECT
  *
FROM
  t3;
```

Eventually I figured out that `t3` query's `FROM` statement I was joining between two different tables. `UNNEST` ([according to the docs](https://trino.io/docs/current/sql/select.html#unnest)) references relations from the left side of the join (which apparently means the tables in the join need to be the same). So I needed to be joining `t2` on the left side with `UNNEST(t2.y)` rather than `t` on the left side.
```
  t3 (x, y) AS (
    SELECT
      x,
      y_unnested
    FROM
      t2
      CROSS JOIN UNNEST (t2.y) t4 (y_unnested)
  )
```

We should add a cleaned up example of this and make that requirement clearer.

**Comment by findepi** [comment]
are you saying docs' example does not work?

> Taken from this awesome example

was this supposed to be a link?

**Comment by martint** [comment]
> references relations from the left side of the join 

Correct. UNNEST is a syntactic sugar for a LATERAL + table function. That means that the tables and columns visible in the invocation of the table function (UNNEST) need to come from the list of tables that appear syntactically to the left of it in the FROM clause.

**Comment by kasiafi** [comment]
@bitsondatadev where does this example come from? It is not the Trino docs, right?

I tried to fix the example so that it refers to `t2` in UNNEST. It is legal, as each WITH-query sees all the preceding WITH-queries. The correct version would be something like
```
WITH
  t (x, y) AS (
    VALUES
      (1, ARRAY['a', 'b', 'c'])
  ),
  t2 (x, y) AS (
    VALUES
      (1, ARRAY['a', 'b', 'c'])
  ),
  t3 (x, y) AS (
    SELECT
      x,
      y_unnested
    FROM
      t
      CROSS JOIN UNNEST ((SELECT y FROM t2)) t4 (y_unnested)
  )
SELECT
  *
FROM
  t3;
```

It does not work currently, because we don't support subqueries inside UNNEST.
Here is a fix to support subqueries in UNNEST: https://github.com/trinodb/trino/pull/17953

**Comment by bitsondatadev** [comment]
ah, good to know!

Your original example was in Slack (I edited the original message): https://trinodb.slack.com/archives/CFLB9AMBN/p1630242769053100?thread_ts=1630241736.052500&cid=CFLB9AMBN

I adapted it to a specific problem I was having though so it's not exactly the same. I was using that example as a reference which helped a lot. This new one will be even better once it's ready!

I think either way I'll come back and document it and if #17953 isn't completed when I do get around to this then I'll just add another ticket to make sure we update the docs when subqueries get merged. Thanks!

**Comment by bitsondatadev** [comment]
Also, if anyone feels like doing this before I get to it please do and just assign yourself to it! This should be approachable for a Trinewbie.

**Comment by soumya-ghosh** [comment]
@bitsondatadev just to clarify, currently we want to emphasize in the docs that columns in UNNEST references columns from relations from the left side of the join, until the https://github.com/trinodb/trino/pull/17953 is merged.

This is already mentioned in docs -
`UNNEST is normally used with a JOIN, and can reference columns from relations on the left side of the join`

Do you want to add another example to docs along the lines of example discussed above?

**Comment by bitsondatadev** [comment]
Correct, this is mentioned but we should provide an example of when this is done and would work, and an example where there's no table on the left join or no join and explain why these fail.

**Comment by soumya-ghosh** [comment]
@bitsondatadev raised PR - https://github.com/trinodb/trino/pull/18379, let me know if it requires changes.

**Comment by datasundae** [comment]
> Also, if anyone feels like doing this before I get to it please do and just assign yourself to it! This should be approachable for a Trinewbie.

Trinewbie reporting for duty :-). I can make a run at this.

**Comment by datasundae** [comment]
Doesn't appear I can "assign myself" (yet?)
